### ISSUE TITLE
[SSM] How to always get latest version of Secure String in Parameter Store

### ISSUE THREAD

**Issue by ookangzheng** [aws/aws-cdk#9793]
### The Question
My team will automatically update SecureString in ParameterStore via CLI, but in CDK, I couldn't / find any way to get around Cloudformation limitation, or workaround (just work) method to always force get the latest version while using DBInstance.

### Environment

  - **CDK CLI Version:** 1.59.0
  - **Module Version:** 1.59.0
  - **Node.js Version:** 14.8.0
  - **OS:** OSX Catalina
  - **Language (Version):** TypeScript (3.9.7)

### Other information
Sample code:

```typescript
const sqlInstance = new DatabaseInstance(this, `Sql-${opt.stage}`, {
      engine: DatabaseInstanceEngine.POSTGRES,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.SMALL),
      masterUsername: SQL_USER_NAME,
      masterUserPassword: ssm.StringParameter.valueForSecureStringParameter(this, "${opt.stage}/SQL_PASSWORD", 
"<Hope I can always get latest version number....>"),
})
```

**Comment by jogold** [comment]
You currently need to use a custom resource to achieve this, have a look at https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/custom-resources#get-the-latest-version-of-a-secure-ssm-parameter

**Comment by ookangzheng** [comment]
I try with this code, but somehow it stuck at DatabaseInstance creation process.

**CDK code**
```typescript
const lambdaRoleStatements: iam.PolicyStatement[] = [
      new iam.PolicyStatement({
        actions: ["ssm:GetParameter", "lambda:InvokeFunction"],
        resources: ["*"]
      })
    ];

    const lambdaExeRole = new iam.Role(this, `ExeRole`, {
      roleName: `BobLambdaExeRole`,
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        {
          managedPolicyArn: "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
        }
      ],
      inlinePolicies: {
        InlinePolicy: new iam.PolicyDocument({
          statements: lambdaRoleStatements
        })
      }
    });

    const getSSMVersionLambda = new lambda.SingletonFunction(this, "Singleton", {
      uuid: "f7d4f730-4ee1-11e8-9c2d-fa7ae01bbebc",
      code: lambda.AssetCode.fromAsset(path.join(__dirname, "lambda")),
      role: lambdaExeRole,
      handler: "get_ssm_version.handler",
      timeout: cdk.Duration.seconds(30),
      runtime: lambda.Runtime.PYTHON_3_8,
      retryAttempts: 2
    });

    const cfnArn = new cfn.CfnCustomResource(this, "CfnGetLambdaArnResource", {
      serviceToken: getSSMVersionLambda.functionArn
    });

    const sqlInstance = new DatabaseInstance(this, `Sql-${opt.stage}`, {
      engine: DatabaseInstanceEngine.POSTGRES,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.SMALL),
      masterUsername: SQL_USER_NAME,
      masterUserPassword: cdk.SecretValue.ssmSecure(
        SQL_PASS_SSM_NAME,
        cfnArn.getAtt("Version").toString()
      ),
      databaseName: "test",
      deletionProtection: false,
      vpc,
      vpcPlacement: {
        subnetType: ec2.SubnetType.PRIVATE
      }
    });
```


**Lambda code**
```python
import boto3
ssm_client = boto3.client('ssm')

def handler(event, context):
    resp = ssm_client.get_parameter(
      Name = '/xxx/dev/xxx_SQL_PASSWORD',
      WithDecryption = True
    )
    print("/xxx/dev/xxx_SQL_PASSWORD latest version: ")
    print(resp['Parameter']['Version'])

    return resp['Parameter']
```

**Comment by ookangzheng** [comment]
Thanks a lot.

**Comment by fengkaijia** [comment]
Can we reopen this so that the AWS team knows currently there's only a workaround but no solution yet?

**Comment by jmm** [comment]
As well as getting the latest version, getting a version by label would be helpful.

The custom resource approach seems like it would complicate making the application deployable without access to the AWS account -- e.g. deploying via a CDK Pipeline by virtue of having cached account context data.

It seems like it could be helpful for CDK to have a context method to retrieve a latest / labeled version of a parameter.

Short of that, perhaps it would work to use the SDK to get the version and manually cache it in context and read it out of context.

**Comment by spyoungtech** [comment]
It would be nice if the version were not required to use a secure string parameter with the [Secret](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.Secret.html) construct because it's not required for [the cloudformation definition](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-secret.html).

**Comment by adikari** [comment]
can we reopen this or at least provide a thought process around why the version is mandatory. I would like want my CDK to always pull the latest version of the secret if the version is not specified.

**Comment by dblackdblack** [comment]
I've asked our AWS TAM to re-open this issue.

**Comment by MrArnoldPalmer** [comment]
Reopening as I'm not sure why this was closed by the PR in guardian/cdk to begin with. I haven't worked on SSM in awhile but we should revisit this and see what guidance we can provide when related to using secure SSM parameters. @dblackdblack is your use case provided by raw cloudformation currently? If so what does the reference to the value look like?

**Comment by rix0rrr** [comment]
CloudFormation has since added support for "latest" versions of Secure SSM parameters.

```ts
SecretValue.ssmSecure('param-name')
```

**Comment by adikari** [comment]
@rix0rrr the reported issue is related to CDK. We get that it's supported in Cloudformation. How, do we achieve this in CDK? Why is this issue closed without a reasonable solution?

**Comment by adikari** [comment]
@dblackdblack you can do following:

```
import { SecretValue } from 'aws-cdk-lib';
SecretValue.ssmSecure('key');
```

**Comment by Mindful** [comment]
```
 SecretValue.ssmSecure(`key)` 
```

Is supported only for a very small number of resources, according to [this](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html#template-parameters-dynamic-patterns-resources). At the very least I can confirm it's not supported for Cognito.

The CustomResource workaround linked earlier in this thread works, but using it means that `cdk diff` always comes back with a diff because the physicalResourceId has changed, even if nothing else has changed. 

Basically it seems like there still isn't a good solution to this problem.
